src/diffusers/models/autoencoders/dc_ae.py

```diff
@@ -0,0 +1,1126 @@
+# Copyright 2024 MIT, Tsinghua University, NVIDIA CORPORATION and The HuggingFace Team.
+# All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Optional, Callable
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.nn import BatchNorm2d
+from huggingface_hub import PyTorchModelHubMixin
+
+from ...configuration_utils import ConfigMixin, register_to_config
+from ..modeling_utils import ModelMixin
+
+from ..activations import get_activation
+from ..normalization import RMSNorm2d
+
+
+__all__ = ["DCAE", "dc_ae_f32c32", "dc_ae_f64c128", "dc_ae_f128c512"]
+
+
+def val2tuple(x: list | tuple | Any, min_len: int = 1) -> tuple:
+    x = list(x) if isinstance(x, (list, tuple)) else [x]
+    # repeat elements if necessary
+    if len(x) > 0:
+        x.extend([x[-1] for _ in range(min_len - len(x))])
+    return tuple(x)
+
+
+def get_same_padding(kernel_size: int | tuple[int, ...]) -> int | tuple[int, ...]:
+    if isinstance(kernel_size, tuple):
+        return tuple([get_same_padding(ks) for ks in kernel_size])
+    else:
+        assert kernel_size % 2 > 0, "kernel size should be odd number"
+        return kernel_size // 2
+
+
+def build_norm(name: Optional[str]="bn2d", num_features: Optional[int]=None) -> Optional[nn.Module]:
+    if name is None:
+        norm = None
+    elif name == "rms2d":
+        norm = RMSNorm2d(normalized_shape=num_features)
+    elif name == "bn2d":
+        norm = BatchNorm2d(num_features=num_features)
+    else:
+        raise ValueError(f"norm {name} is not supported")
+    return norm
+
+
+class ConvLayer(nn.Module):
```

Comment:
This is a wrapper that has a Conv2d layer, a norm layer, and an activation function. I think we should separate them into atomic pieces and make the following configurable from the model initialization:

* `norm_layer`
* `activation_fn`

Based on the values provided by the user, we can propagate them to decide which layer to use. 

Here is an example of `norm_layer`:
https://github.com/huggingface/diffusers/blob/9a92b8177cb3f8bf4b095fff55da3b45a3607960/src/diffusers/models/transformers/transformer_2d.py#L103

`activation_fn`:
https://github.com/huggingface/diffusers/blob/9a92b8177cb3f8bf4b095fff55da3b45a3607960/src/diffusers/models/transformers/transformer_2d.py#L85

LMK if it makes sense?

Reply:
I have a concern here... There might be issues loading the pretrained weights due to the wrapper class. Also, the pretrained weights have been uploaded to hugging face and used in our original repo.

Reply:
Is it possible for us to convert these weights each time we load models? Since we are not willing to store the weights somewhere else, because we want to track download statistics.

Reply:
We can host them separately like https://huggingface.co/stabilityai/stable-diffusion-3-medium-diffusers does. 

State dict conversion during loading isn't something that we do when `from_pretrained()` is called. 

Ccing @yiyixuxu @DN6 here too. What do you think?

Reply:
We had a discussion and decided to also upload the converted models to hugging face.

Reply:
That makes sense then. I guess we have reached a resolution?

Reply:
Yes, I think so. Thanks for the discussion.